pubsub/consumer.go

```diff
@@ -94,82 +82,105 @@ func (c *Consumer[Request, Response]) StreamName() string {
 	return c.redisStream
 }
 
-func (c *Consumer[Request, Response]) heartBeatKey() string {
-	return heartBeatKey(c.id)
-}
-
-// deleteHeartBeat deletes the heartbeat to indicate it is being shut down.
-func (c *Consumer[Request, Response]) deleteHeartBeat(ctx context.Context) {
-	if err := c.client.Del(ctx, c.heartBeatKey()).Err(); err != nil {
-		l := log.Info
-		if ctx.Err() != nil {
-			l = log.Error
-		}
-		l("Deleting heardbeat", "consumer", c.id, "error", err)
-	}
-}
-
-// heartBeat updates the heartBeat key indicating aliveness.
-func (c *Consumer[Request, Response]) heartBeat(ctx context.Context) {
-	if err := c.client.Set(ctx, c.heartBeatKey(), time.Now().UnixMilli(), 2*c.cfg.KeepAliveTimeout).Err(); err != nil {
-		l := log.Info
-		if ctx.Err() != nil {
-			l = log.Error
-		}
-		l("Updating heardbeat", "consumer", c.id, "error", err)
-	}
-}
-
 // Consumer first checks it there exists pending message that is claimed by
 // unresponsive consumer, if not then reads from the stream.
-func (c *Consumer[Request, Response]) Consume(ctx context.Context) (*Message[Request], error) {
-	res, err := c.client.XReadGroup(ctx, &redis.XReadGroupArgs{
+func (c *Consumer[Request, Response]) Consume(ctx context.Context) (*Message[Request], chan struct{}, error) {
+	// First try to XAUTOCLAIM, this prioritizes processing PEL messages
+	// that have been waiting for more than IdletimeToAutoclaim duration
+	messages, _, err := c.client.XAutoClaim(ctx, &redis.XAutoClaimArgs{
 		Group:    c.redisGroup,
 		Consumer: c.id,
-		// Receive only messages that were never delivered to any other consumer,
-		// that is, only new messages.
-		Streams: []string{c.redisStream, ">"},
-		Count:   1,
-		Block:   time.Millisecond, // 0 seems to block the read instead of immediately returning
+		MinIdle:  c.cfg.IdletimeToAutoclaim, // Minimum idle time for messages to claim (in milliseconds)
+		Stream:   c.redisStream,
+		Start:    "0",
+		Count:    5, // Try looking for 50 entries in PEL, this assumes there are a maximum of 50 consumers in this redisGroup
```

Comment:
I don't think that setting count > 1 will work for us, I've added details under original discussion: https://github.com/OffchainLabs/nitro/pull/2581#discussion_r1732791331